File: ios/FinvuModule.swift
import ExpoModulesCore
import FinvuSDK

class FinvuClientConfig: FinvuConfig {
    var finvuEndpoint: URL
    var certificatePins: [String]?
    
    public init(finvuEndpoint: URL, certificatePins: [String]?) {
        self.finvuEndpoint = finvuEndpoint
        self.certificatePins = certificatePins ?? []
    }
}

public class FinvuModule: Module {
    private let sdkInstance = FinvuManager.shared

    public func definition() -> ModuleDefinition {
        Name("Finvu")
        
        Events("onConnectionStatusChange", "onLoginOtpReceived", "onLoginOtpVerified")

        Function("initializeWith") { (config: [String: Any]) -> String in
            do {
                guard let finvuEndpointString = config["finvuEndpoint"] as? String,
                      let finvuEndpoint = URL(string: finvuEndpointString) else {
                    throw NSError(domain: "FinvuModule", code: -1, userInfo: [NSLocalizedDescriptionKey: "finvuEndpoint is required and must be a valid URL"])
                }
                
                let certificatePins = (config["certificatePins"] as? [String])
                let finvuClientConfig = FinvuClientConfig(finvuEndpoint: finvuEndpoint, certificatePins: certificatePins)
                
                sdkInstance.initializeWith(config: finvuClientConfig)
                return "Initialized successfully"
            } catch {
                print(error)
                throw NSError(domain: "FinvuModule", code: -1, userInfo: [NSLocalizedDescriptionKey: "INITIALIZATION_ERROR"])
            }
        }

        AsyncFunction("connect") { (promise: Promise) in
            sdkInstance.connect { error in
                DispatchQueue.main.async {
                    if let error = error as NSError? {
                        // Convert NSError to React Native error
                        let errorCode: String
                        // Map NSError codes to custom error codes if needed
                        switch error.code {
                        case 1001: // Replace with specific error codes if applicable
                            errorCode = "SPECIFIC_ERROR_CODE"
                        default:
                            errorCode = "UNKNOWN_ERROR"
                        }
                        self.sendEvent("onConnectionStatusChange", ["status": errorCode])
                        promise.reject(errorCode, error.localizedDescription)
                    } else {
                        // No error means success
                        self.sendEvent("onConnectionStatusChange", ["status": "Connected successfully"])
                        promise.resolve(["status": "Connected successfully"])
                    }
                }
            }
        }

        AsyncFunction("loginWithUsernameOrMobileNumber") { (username: String, mobileNumber: String, consentHandleId: String, promise: Promise) in
            sdkInstance.loginWith(username: username, mobileNumber: mobileNumber, consentHandleId: consentHandleId) { result, error in
                DispatchQueue.main.async {
                    if let error = error as NSError? {
                        // Convert NSError to React Native error
                        let errorCode: String
                        // Map NSError codes to custom error codes if needed
                        switch error.code {
                        case 1001: // Replace with specific error codes if applicable
                            errorCode = "SPECIFIC_ERROR_CODE"
                        default:
                            errorCode = "UNKNOWN_ERROR"
                        }
                        promise.reject(errorCode, error.localizedDescription)
                    } else if let result = result {
                        // Handle successful login
                        let reference = result.reference
                        promise.resolve(["reference": reference])
                    } else {
                        // Handle case where there is no result and no error
                        promise.reject("UNKNOWN_ERROR", "An unknown error occurred.")
                    }
                }
            }
        }

                                  
        AsyncFunction("verifyLoginOtp") { (otp: String, otpReference: String, promise: Promise) in
            sdkInstance.verifyLoginOtp(otp: otp, otpReference: otpReference) { result, error in
                DispatchQueue.main.async {
                    if let error = error as NSError? {
                        // Convert NSError to React Native error
                        let errorCode: String
                        // Map NSError codes to custom error codes if needed
                        switch error.code {
                        case 1001: // Replace with specific error codes if applicable
                            errorCode = "SPECIFIC_ERROR_CODE"
                        default:
                            errorCode = "UNKNOWN_ERROR"
                        }
                        promise.reject(errorCode, error.localizedDescription)
                    } else if let result = result {
                        // Handle successful OTP verification
                        let userId = result.userId
                        promise.resolve(["userId": userId])
                    } else {
                        // Handle case where there is no result and no error
                        promise.reject("UNKNOWN_ERROR", "An unknown error occurred.")
                    }
                }
            }
        }
    }
}

File: android/src/main/java/expo/modules/finvu/FinvuModule.kt
package expo.modules.finvu

import com.finvu.android.FinvuManager
import com.finvu.android.publicInterface.FinvuErrorCode
import com.finvu.android.publicInterface.FinvuException
import com.finvu.android.utils.FinvuConfig
import expo.modules.kotlin.Promise
import expo.modules.kotlin.modules.Module
import expo.modules.kotlin.modules.ModuleDefinition
import java.net.URL

class FinvuModule : Module() {

  data class FinvuClientConfig(
    override val finvuEndpoint: String,
    override val certificatePins: List<String>?
  ) : FinvuConfig

  private val sdkInstance: FinvuManager = FinvuManager.shared

  override fun definition() = ModuleDefinition {
    Name("Finvu")
    Constants(
      // Add any constants here if needed
    )

    Events("onConnectionStatusChange", "onLoginOtpReceived", "onLoginOtpVerified")

    Function("initializeWith") { config: Map<String, Any> ->
      try {
        val finvuEndpoint = config["finvuEndpoint"] as? String ?: throw IllegalArgumentException("finvuEndpoint is required")
        val certificatePins = (config["certificatePins"] as? List<*>)?.map { it.toString() }

        val finvuClientConfig = FinvuClientConfig(finvuEndpoint, certificatePins)
        sdkInstance.initializeWith(finvuClientConfig)
        "Initialized successfully"
      } catch (e: Exception) {
        e.printStackTrace()
        throw RuntimeException("INITIALIZATION_ERROR", e)
      }
    }

    AsyncFunction("connect") {
      sdkInstance.connect { result ->
          if (result.isSuccess) {
            sendEvent("onConnectionStatusChange", mapOf("status" to "Connected successfully"))
          } else {
            val exception = result.exceptionOrNull() as? FinvuException
            if (exception != null) {
              val errorCode = when (exception.code) {
                FinvuErrorCode.SSL_PINNING_FAILURE_ERROR.code -> "SSL_PINNING_FAILURE_ERROR"
                else -> "UNKNOWN_ERROR"
              }
              sendEvent("onConnectionStatusChange", mapOf("status" to errorCode))
              throw RuntimeException(errorCode)
            } else {
              sendEvent("onConnectionStatusChange", mapOf("status" to "UNKNOWN_ERROR"))
              throw RuntimeException("UNKNOWN_ERROR")
            }
          }
      }
    }

    AsyncFunction("loginWithUsernameOrMobileNumber") { username: String, mobileNumber: String, consentHandleId: String,promise: Promise ->
      try {
        sdkInstance.loginWithUsernameOrMobileNumber(username, mobileNumber, consentHandleId) { result ->
            if (result.isSuccess) {
              val reference = result.getOrNull()?.reference
              promise.resolve(mapOf("reference" to reference))
            } else {
              val exception = result.exceptionOrNull() as? FinvuException
              val errorCode = exception?.code ?: "UNKNOWN_ERROR"
              promise.reject(errorCode.toString(), exception?.message, null)
              //sendEvent("onLoginOtpReceived", mapOf("status" to errorCode))
              //throw RuntimeException(errorCode.toString())
            }
          }
      } catch (e: Exception) {
        e.printStackTrace()
        throw RuntimeException("CONNECT_ERROR", e)
      }
    }

    AsyncFunction("verifyLoginOtp") { otp: String, otpReference: String, promise: Promise ->
      try {
        sdkInstance.verifyLoginOtp(otp, otpReference) { result ->
            if (result.isSuccess) {
              val userId = result.getOrNull()?.userId
              promise.resolve(mapOf("userId" to userId))
            } else {
              val exception = result.exceptionOrNull() as? FinvuException
              val errorCode = exception?.code ?: "UNKNOWN_ERROR"
              promise.reject(errorCode.toString(), exception?.message,null)
              throw RuntimeException(errorCode.toString())
            }
        }
      } catch (e: Exception) {
        e.printStackTrace()
        throw RuntimeException("CONNECT_ERROR", e)
      }
    }
  }
}